// @cloudflare/ai/dist/tasks/image-classification.js
import { Tensor, TensorType } from "../tensor";
import { resnetLabels } from "./data/labels";
export class AiImageClassification {
    constructor(inputs, modelSettings) {
        this.schema = {
            input: {
                oneOf: [
                    { type: "string", format: "binary" },
                    {
                        type: "object",
                        properties: {
                            image: {
                                type: "array",
                                items: {
                                    type: "number",
                                },
                            },
                        },
                    },
                ],
            },
            output: {
                type: "array",
                contentType: "application/json",
                items: {
                    type: "object",
                    properties: {
                        score: {
                            type: "number",
                        },
                        label: {
                            type: "string",
                        },
                    },
                },
            },
        };
        this.inputs = inputs;
        this.modelSettings = modelSettings;
    }
    preProcessing() {
        this.preProcessedInputs = this.inputs;
    }
    generateTensors() {
        this.tensors = [
            new Tensor(TensorType.Uint8, this.preProcessedInputs.image, {
                shape: [1, this.preProcessedInputs.image.length],
                name: "input",
            }),
        ];
    }
    postProcessing(response) {
        const labels = [];
        const scores = response.output.value[0];
        for (var s in scores)
            labels.push({ label: resnetLabels[s], score: scores[s] });
        labels.sort((a, b) => {
            return b.score - a.score;
        });
        this.postProcessedOutputs = labels.slice(0, 5);
    }
}
