// @cloudflare/ai/dist/tasks/object-detection.js
import { Tensor, TensorType } from "../tensor";
export class AiObjectDetection {
    constructor(inputs, modelSettings) {
        this.schema = {
            input: {
                oneOf: [
                    { type: "string", format: "binary" },
                    {
                        type: "object",
                        properties: {
                            image: {
                                type: "array",
                                items: {
                                    type: "number",
                                },
                            },
                        },
                    },
                ],
            },
            output: {
                type: "array",
                contentType: "application/json",
                items: {
                    type: "object",
                    properties: {
                        score: {
                            type: "number",
                        },
                        label: {
                            type: "string",
                        },
                        box: {
                            type: "object",
                            properties: {
                                xmin: {
                                    type: "number",
                                },
                                ymin: {
                                    type: "number",
                                },
                                xmax: {
                                    type: "number",
                                },
                                ymax: {
                                    type: "number",
                                },
                            },
                        },
                    },
                },
            },
        };
        this.inputs = inputs;
        this.modelSettings = modelSettings;
    }
    preProcessing() {
        this.preProcessedInputs = this.inputs;
    }
    generateTensors() {
        this.tensors = [
            new Tensor(TensorType.Uint8, this.preProcessedInputs.image, {
                shape: [1, this.preProcessedInputs.image.length],
                name: "input",
            }),
        ];
    }
    postProcessing(response) {
        const scores = response.scores.value[0].map((score, i) => {
            return {
                score: score,
                label: response.label_names.value[response.labels.value[0][i]],
                box: {
                    xmin: response.boxes.value[0][i][0],
                    ymin: response.boxes.value[0][i][1],
                    xmax: response.boxes.value[0][i][2],
                    ymax: response.boxes.value[0][i][3],
                },
            };
        });
        this.postProcessedOutputs = scores;
    }
}
