// @cloudflare/ai/dist/tasks/text-classification.js
import { Tensor, TensorType } from "../tensor";
export class AiTextClassification {
    constructor(inputs, modelSettings) {
        this.schema = {
            input: {
                type: "object",
                properties: {
                    text: {
                        type: "string",
                    },
                },
                required: ["text"],
            },
            output: {
                type: "array",
                contentType: "application/json",
                items: {
                    type: "object",
                    properties: {
                        score: {
                            type: "number",
                        },
                        label: {
                            type: "string",
                        },
                    },
                },
            },
        };
        this.inputs = inputs;
        this.modelSettings = modelSettings;
    }
    preProcessing() {
        this.preProcessedInputs = this.inputs;
    }
    generateTensors() {
        this.tensors = [
            new Tensor(TensorType.String, [this.preProcessedInputs.text], {
                shape: [1],
                name: "input_text",
            }),
        ];
    }
    postProcessing(response) {
        this.postProcessedOutputs = [
            {
                label: "NEGATIVE",
                score: response.logits.value[0][0],
            },
            {
                label: "POSITIVE",
                score: response.logits.value[0][1],
            },
        ];
    }
}
