// @cloudflare/ai/dist/tasks/text-to-image.js
import { Tensor, TensorType } from "../tensor";
export class AiTextToImage {
    constructor(inputs, modelSettings) {
        this.schema = {
            input: {
                type: "object",
                properties: {
                    prompt: {
                        type: "string",
                    },
                    num_steps: {
                        type: "integer",
                        default: 20,
                        maximum: 20,
                    },
                },
                required: ["prompt"],
            },
            output: {
                type: "string",
                contentType: "image/png",
                format: "binary",
            },
        };
        this.inputs = inputs;
        this.modelSettings = modelSettings;
    }
    preProcessing() {
        this.preProcessedInputs = this.inputs;
    }
    generateTensors() {
        this.tensors = [
            new Tensor(TensorType.String, [this.preProcessedInputs.prompt], {
                shape: [1],
                name: "input_text",
            }),
            new Tensor(TensorType.Int32, [this.preProcessedInputs.num_steps], {
                shape: [1],
                name: "num_steps",
            }),
        ];
    }
    postProcessing(response) {
        this.postProcessedOutputs = new Uint8Array(response.output_image.value);
    }
}
